compiler/rustc_typeck/src/expr_use_visitor.rs

```diff
@@ -608,6 +662,41 @@ impl<'a, 'tcx> ExprUseVisitor<'a, 'tcx> {
             ty::Closure(..) | ty::Generator(..)
         );
 
+        if let Some(fake_reads) = self.mc.typeck_results.closure_fake_reads.get(&closure_def_id) {
+            for (fake_read, cause) in fake_reads.iter() {
+                match fake_read.base {
+                    PlaceBase::Upvar(upvar_id) => {
+                        if upvars.map_or(body_owner_is_closure, |upvars| {
+                            !upvars.contains_key(&upvar_id.var_path.hir_id)
+                        }) {
+                            // The nested closure might be fake reading the current (enclosing) closure's local variables.
+                            // We only want to fake read the fake read present in the nested closure that are not part of
+                            // the enclosing closure's local variables.
+                            //
+                            // ```rust,ignore(cannot-test-this-because-pseduo-code)
+                            // let v1 = (0, 1);
+                            // let c = || { // fake reads: v1
+                            //    let v2 = (0, 1);
+                            //    let e = || { // fake reads: v1, v2
+                            //       let (_, t1) = v1;
+                            //       let (_, t2) = v2;
+                            //    }
+                            // }
+                            // ```
```

Comment:
This code example looks right to me, but I can't quite tell what the purpose of this check is -- I would sort of expect this to just "fall out' from other bits of the code working correctly. That is, I would expect that the `closure_fake_reads` field on the typeck results for `c` would only have `v1`, and the same field for `e` would have `v1` and `v2`. Am I wrong about this?

It might help if you clarified whether you expect this `continue` to fire when visiting the outermost body here (that contains `let c = ...`) or when visiting the body of `c` (which contains `let e = ...`).

Reply:
The fake read of `v1` in `c` only happens here in the code. It's similar idea to how captures from the nest closures are added to the outermost closures here: https://github.com/rust-lang/rust/blob/bfcecbd0dc429014395062ab0df4d2306cc76766/compiler/rustc_typeck/src/expr_use_visitor.rs#L700

Reply:
Side note that I personally find `map_or` to be very hard to read. I tend to prefer `foo.map(...).unwrap_or(...)` -- map-or requires me to read the fallback case before the function, out of order with what happens at runtime, and it's confusing :/

Reply:
OK, so, I think this is what is going on:

* We read the upvar list of the "inner" closure.
* For each upvar, we check if it is *also* an upvar for the current closure. 
* If so, we generate a fake read?

I think I find this confusing also for the other code down below. I think what I expect is:

* We go through the fake reads from the inner closure, each of which is a fake read of a captured path P.
* I think those captured paths P are expressed in terms of the closure creator (i.e., us), so no translation is needed, we just do a `delegate.fake_read(captured_path)`

But if translation is needed, then maybe I would expect a helper that says "is this upvar a local variable here?", and then we could just use that helper here and below.

Does that make sense? Am I missing something? (cc @arora-aman, I think the code that @roxelo cited is code you wrote? Or maybe I'm misremembering)

Reply:
That is kind of what's happening. There is a code sample but it is at function-doc level

https://github.com/rust-lang/rust/blob/master/compiler/rustc_typeck/src/expr_use_visitor.rs#L589-L595

Maybe we should bring it within the function

> I think those captured paths P are expressed in terms of the closure creator (i.e., us), so no translation is needed, we just do a delegate.fake_read(captured_path) 

I don't think this is correct, the fake reads that are stored within an inner/nested closure are (and should be) from the prespective of that nested closure. When we take this and convert it from the prespective of the enclosing/outer closure we'd have to convert the Place to say so.
For reference see: https://github.com/rust-lang/rust/blob/master/compiler/rustc_typeck/src/expr_use_visitor.rs#L624-L631


I like the idea behind the having a helper because this was confusing earlier aswell.